### ISSUE TITLE
Enhancement: [no-unused-vars] catch unused key in anonymous object type destructure

### ISSUE THREAD

**Issue by qpwo** [typescript-eslint/typescript-eslint#6298]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unused-vars

### Description

If you have a destructure with an inline (i.e. anonymous) object type then you should get a warning if you don't use one of the properties/keys. This encourages removal of unnecessary arguments which cascades up the call tree and can greatly simplify code.

I don't think that catching unused properties on named types is worth the trouble and confusion.

### Fail

```typescript
// function signature (more important case)

function calculate({x, y}: {x: number, y: number, scalar: number}) {
  return x * y
}

// assignment (less important)
const {a, b}: {a: string, b: string, c: string} = o
```

### Pass

```typescript
function calculate({x, y}: {x: number, y: number}) {
  return x * y
}

const {a, b, c}: {a: string, b: string, c: string} = o

type T = {x: number, y: number}
function f({x}: T) {
  return x
}
```

**Comment by bradzacher** [comment]
I'm not sure how generally this rule would apply to most of the community.
A lot of libraries define non-anonymous types for the object arguments as it allows consumers to reference those types - or more generally it's rare to create an unordered parameter set with an object (and more common to be accepting an object argument that isn't destructured).

However that case isn't easy to cover because a type alias may have more than one usage - thus suggesting that a specific prop is unused may be wrong. That being said - it's entirely possible to check non-exported type aliases that are only used once. Though again that's pretty limited in scope.

The one time that I have really seen this as a common pattern across a codebase is in react functional components. That specific usecase should be covered by the external rule [`react/no-unused-prop-types`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unused-prop-types.md).

To be clear - I definitely see value in the concept of this rule - removing dead code is a big A+ from me. What I struggle to see is how much dead code this would actually catch in most codebases.

It would probably be worth using [sourcegraph](https://sourcegraph.com/search) or [github codesearch](https://cs.github.com/) to try and get a rough idea of if this pattern is common outside react components. If it's common then this rule would be a good addition, if not then it would not be a good addition here.

**Comment by bradzacher** [comment]
I think the coding style that enforces is very, very strict and not something most people would want to do. It's a lot more cumbersome to do and much harder to read.

Ultimately there's a large amount of complexity in such a check - and there's a lot of cases that can't be handled effectively without type information (which won't *ever* be added to `no-unused-vars`).

After 6 months of being open there also hasn't been any community engagement on this issue - which suggests this isn't something the community finds value in.

For all of the above reasons, I'm going to close this proposal as rejected.
